File: app/planilha.py
"""Leitura e gravação de planilhas no formato exportado pelo Effort.

Observação: o Effort exporta arquivos com extensão .xls que na verdade são
.xlsx. Por isso o formato é detectado pelo conteúdo, não pela extensão.
"""
import io

from openpyxl import Workbook, load_workbook
from openpyxl.styles import Alignment, Font, PatternFill
from openpyxl.utils import get_column_letter

import config
from app.utils import sem_acentos, texto

ASSINATURA_XLSX = b"PK"
ASSINATURA_XLS = b"\xD0\xCF\x11\xE0"


class ErroPlanilha(Exception):
    pass


def _normalizar_titulo(titulo):
    return " ".join(sem_acentos(titulo).lower().split())


_TITULO_PARA_CAMPO = {
    _normalizar_titulo(titulo): campo for campo, titulo in config.COLUNAS.items()
}


def _linhas_xlsx(conteudo):
    livro = load_workbook(io.BytesIO(conteudo), read_only=True, data_only=True)
    aba = livro.active
    for linha in aba.iter_rows(values_only=True):
        yield list(linha)
    livro.close()


def _linhas_xls(conteudo):
    try:
        import xlrd
    except ImportError as erro:
        raise ErroPlanilha("Para ler .xls antigos instale o pacote xlrd.") from erro
    livro = xlrd.open_workbook(file_contents=conteudo)
    aba = livro.sheet_by_index(0)
    for i in range(aba.nrows):
        yield aba.row_values(i)


def ler(conteudo):
    """Recebe os bytes do arquivo e devolve uma lista de dicionários."""
    if conteudo.startswith(ASSINATURA_XLSX):
        linhas = _linhas_xlsx(conteudo)
    elif conteudo.startswith(ASSINATURA_XLS):
        linhas = _linhas_xls(conteudo)
    else:
        raise ErroPlanilha("O arquivo não parece ser uma planilha do Excel.")

    try:
        cabecalho = next(linhas)
    except StopIteration:
        raise ErroPlanilha("A planilha está vazia.")

    mapa = {}  # índice da coluna -> campo interno
    for indice, titulo in enumerate(cabecalho):
        campo = _TITULO_PARA_CAMPO.get(_normalizar_titulo(titulo))
        if campo:
            mapa[indice] = campo

    campos_encontrados = set(mapa.values())
    if "equipamento" not in campos_encontrados or not (
            {"setor_origem", "setor_atual"} & campos_encontrados):
        raise ErroPlanilha(
            "Colunas obrigatórias não encontradas. A planilha precisa ter "
            "pelo menos 'Equipamento' e 'Setor Origem' ou 'Setor Atual'.")

    itens = []
    for linha in linhas:
        item = {campo: texto(linha[i]) if i < len(linha) else ""
                for i, campo in mapa.items()}
        if any(item.values()):
            itens.append(item)
    return itens


def gerar(itens):
    """Gera um .xlsx com as mesmas colunas da importação. Devolve BytesIO."""
    livro = Workbook()
    aba = livro.active
    aba.title = config.NOME_ABA_EXPORTACAO
    campos = list(config.COLUNAS.keys())

    fonte = Font(name="Arial", size=10)
    fonte_titulo = Font(name="Arial", size=10, bold=True, color="FFFFFF")
    fundo_titulo = PatternFill("solid", fgColor="1F4E5A")

    aba.append([config.COLUNAS[c] for c in campos])
    for celula in aba[1]:
        celula.font = fonte_titulo
        celula.fill = fundo_titulo
        celula.alignment = Alignment(vertical="center")

    for item in itens:
        # Tudo como texto para não perder zeros à esquerda (ex.: 062.013158)
        aba.append([texto(item.get(c)) or None for c in campos])

    larguras = {"equipamento": 60, "observacao": 35, "setor_origem": 28,
                "setor_atual": 28, "executante": 30, "motivo": 26}
    for indice, campo in enumerate(campos, start=1):
        letra = get_column_letter(indice)
        aba.column_dimensions[letra].width = larguras.get(campo, 18)
        for celula in aba[letra][1:]:
            celula.font = fonte
            celula.number_format = "@"

    aba.freeze_panes = "A2"
    aba.auto_filter.ref = aba.dimensions

    saida = io.BytesIO()
    livro.save(saida)
    saida.seek(0)
    return saida

File: mobile/js/planilha.js
// Leitura e gravação de planilhas no formato exportado pelo Effort
// (port de app/planilha.py, usando SheetJS no lugar do openpyxl/xlrd).
import * as config from "./config.js";
import { semAcentos, texto } from "./utils.js";

// globalThis (não window) para o mesmo módulo funcionar tanto no navegador
// quanto nos testes em Node (mobile/tests/logica.test.mjs).
const XLSX = globalThis.XLSX;

export class ErroPlanilha extends Error {}

function normalizarTitulo(titulo) {
  return semAcentos(titulo).toLowerCase().split(/\s+/).filter(Boolean).join(" ");
}

function mapaTituloParaCampo(colunas) {
  const mapa = new Map();
  for (const [campo, titulo] of Object.entries(colunas)) {
    mapa.set(normalizarTitulo(titulo), campo);
  }
  return mapa;
}

const TITULO_PARA_CAMPO = mapaTituloParaCampo(config.COLUNAS);
const TITULO_PARA_CAMPO_EFFORT = mapaTituloParaCampo(config.COLUNAS_REFERENCIA_EFFORT);

// Classificações que o próprio app deriva sozinho durante a auditoria
// (comparando setor de origem/atual). A planilha de exemplo carregada na
// primeira abertura do app (dados_padrao/Auditoria.xlsx) não tem garantia de
// que essas classificações ainda estejam corretas, então elas voltam para
// "Pendente" nesse caso — ver `revisarMotivos` em `ler()`.
//
// Isso NÃO deve ser aplicado ao mesclar planilhas exportadas por este próprio
// aplicativo (fluxo normal entre auditores): esses valores já são o resultado
// de uma auditoria real feita em outro celular e devem ser respeitados; do
// contrário a mesclagem os transforma em "Pendente" e a regra de mesclagem
// (um pendente e outro não pendente -> fica o não pendente) descarta o
// trabalho já feito, fazendo o total de equipamentos auditados CAIR depois
// de importar uma planilha — que é exatamente o oposto do esperado.
const MOTIVOS_A_REVISAR_NA_IMPORTACAO = new Set([
  config.MOTIVO_OUTRO_LOCAL,
  ...Object.keys(config.OUTROS_MOTIVOS),
]);

function revisarMotivoImportado(item) {
  if (MOTIVOS_A_REVISAR_NA_IMPORTACAO.has(item.motivo)) {
    item.motivo = config.MOTIVO_PENDENTE;
    item.status = config.STATUS_NAO_CONFORME;
  }
  return item;
}

// Recebe os bytes do arquivo (ArrayBuffer/Uint8Array) e devolve a primeira
// aba como uma lista de linhas (cada linha uma lista de valores).
function linhasDaPrimeiraAba(bytes) {
  const dados = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
  let livro;
  try {
    livro = XLSX.read(dados, { type: "array", cellDates: true });
  } catch (erro) {
    throw new ErroPlanilha("O arquivo não parece ser uma planilha do Excel.");
  }
  const nomeAba = livro.SheetNames[0];
  if (!nomeAba) throw new ErroPlanilha("A planilha está vazia.");
  const aba = livro.Sheets[nomeAba];
  return XLSX.utils.sheet_to_json(aba, { header: 1, raw: false, defval: "" });
}

function lerComMapa(bytes, mapaTitulos) {
  const linhas = linhasDaPrimeiraAba(bytes);
  if (!linhas.length) throw new ErroPlanilha("A planilha está vazia.");

  const [cabecalho, ...resto] = linhas;
  const mapa = new Map(); // índice da coluna -> campo interno
  cabecalho.forEach((titulo, indice) => {
    const campo = mapaTitulos.get(normalizarTitulo(titulo));
    if (campo) mapa.set(indice, campo);
  });

  return { mapa, linhas: resto };
}

// Recebe os bytes do arquivo e devolve uma lista de objetos.
// `revisarMotivos`: true só para a planilha de exemplo (garantirAuditoriaInicial).
// Ao mesclar planilhas de auditores (uso normal), deixe false para não perder
// auditorias já feitas em outro celular — ver MOTIVOS_A_REVISAR_NA_IMPORTACAO acima.
export function ler(bytes, { revisarMotivos = false } = {}) {
  const { mapa, linhas } = lerComMapa(bytes, TITULO_PARA_CAMPO);
  const camposEncontrados = new Set(mapa.values());
  if (!camposEncontrados.has("equipamento") ||
      !(camposEncontrados.has("setor_origem") || camposEncontrados.has("setor_atual"))) {
    throw new ErroPlanilha(
      "Colunas obrigatórias não encontradas. A planilha precisa ter " +
      "pelo menos 'Equipamento' e 'Setor Origem' ou 'Setor Atual'.");
  }

  const itens = [];
  for (const linha of linhas) {
    const item = {};
    for (const [indice, campo] of mapa) item[campo] = texto(linha[indice]);
    if (Object.values(item).some(Boolean)) {
      itens.push(revisarMotivos ? revisarMotivoImportado(item) : item);
    }
  }
  return itens;
}

// Lê a planilha bruta de equipamentos exportada do Effort (ID, TAG, Setor...).
// É um formato bem mais largo; só as colunas de config.COLUNAS_REFERENCIA_EFFORT
// são aproveitadas, o resto é ignorado.
export function lerReferenciaEffort(bytes) {
  const { mapa, linhas } = lerComMapa(bytes, TITULO_PARA_CAMPO_EFFORT);
  if (![...mapa.values()].includes("id_effort")) {
    throw new ErroPlanilha("Coluna 'ID' não encontrada na planilha de equipamentos.");
  }

  const itens = [];
  for (const linha of linhas) {
    const item = {};
    for (const [indice, campo] of mapa) item[campo] = texto(linha[indice]);
    if (item.id_effort) itens.push(item);
  }
  return itens;
}

// Gera um .xlsx com as mesmas colunas da importação. Devolve um Blob.
export function gerar(itens) {
  const campos = Object.keys(config.COLUNAS);
  const linhas = [campos.map((c) => config.COLUNAS[c])];
  for (const item of itens) {
    linhas.push(campos.map((c) => texto(item[c]) || ""));
  }

  const aba = XLSX.utils.aoa_to_sheet(linhas);
  const larguras = { equipamento: 60, observacao: 35, setor_origem: 28,
    setor_atual: 28, executante: 30, motivo: 26 };
  aba["!cols"] = campos.map((c) => ({ wch: larguras[c] || 18 }));

  const livro = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(livro, aba, config.NOME_ABA_EXPORTACAO);
  const saida = XLSX.write(livro, { type: "array", bookType: "xlsx" });
  return new Blob([saida], {
    type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  });
}
